### PR TITLE
[Spot] Fix spot pending status

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR fixes a problem caused by #1636, where the spot job not showing up in the `sky spot queue` if the spot job is in PENDING status. This is found when running the test in #2041.

Before this PR:
Only 1 pending spot job will show up in the `sky spot queue`, other later pending jobs will only appear when it starts running.

After this PR:
All the pending spot jobs will also appear in the `sky spot queue` with PENDING status.

Tested (run the relevant ones):

- [x] Any manual or new tests for this PR (please specify below)
  - [x] Submit more than 16 spot jobs, and all the spot jobs should be in the correct status.
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/backend_utils.py [hunk 1]
```diff
@@ -1538,21 +1538,28 @@ def _get_tpu_vm_pod_ips(ray_config: Dict[str, Any],
...
-    use_cached_head_ip: bool = True,
+    use_cached_head_ip: Optional[bool] = True,
     max_attempts: int = 1,
 ) -> str:
-    """Returns the ip of the head node."""
+    """Returns the ip of the head node.
+
+    use_cached_head_ip: If True, use the cached head ip if it exists. If False,
+        query the head ip from the cloud provider. If None, use the cached head
+        ip if it exists, otherwise query the head ip from the cloud provider.
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1548]: Add Args: and + other args?

File: sky/backends/cloud_vm_ray_backend.py [hunk 2]
```diff
@@ -1821,7 +1810,7 @@ def _ensure_cluster_ray_started(self, handle: 'CloudVmRayResourceHandle',
             # At this state, an erroneous cluster may not have cached
             # handle.head_ip (global_user_state.add_or_update_cluster(...,
             # ready=True)).
-            use_cached_head_ip=False)
+            use_cached_head_ip=None)
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:1813]: There are a couple places in this PR where previously we always query the cloud for IPs, but with this PR we always try to use cached IPs first. Why is that?

File: sky/backends/cloud_vm_ray_backend.py [hunk 3]
```diff
@@ -3575,15 +3568,21 @@ def run_on_head(
...
         NON_INTERACTIVE,
         under_remote_workdir: bool = False,
         require_outputs: bool = False,
         separate_stderr: bool = False,
         **kwargs,
     ) -> Union[int, Tuple[int, str, str]]:
-        """Runs 'cmd' on the cluster's head node."""
+        """Runs 'cmd' on the cluster's head node.
+
+        use_cached_head_ip: If True, use the cached head IP address. If False,
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3581]: Can we add other Args too? (I know, code gardening...)

File: sky/backends/cloud_vm_ray_backend.py [hunk 4]
```diff
@@ -3946,7 +3956,9 @@ def _execute_task_one_node(self, handle: CloudVmRayResourceHandle,
                                 codegen.build(),
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:3956]: How about
- `codegen.build() + ' && ' + self._maybe_set_spot_task_to_pending_code(...)`?
- add a comment on why inside the generated code, we call spot_state.set_pend... only after the `codegen` command has been run?

File: sky/backends/cloud_vm_ray_backend.py [hunk 5]
```diff
@@ -2790,6 +2780,9 @@ def _exec_code_on_head(
             mkdir_code = (f'{cd} && mkdir -p {remote_log_dir} && '
                           f'touch {remote_log_path}')
             code = job_lib.JobLibCodeGen.queue_job(job_id, job_submit_cmd)
+            if extra_cmd is not None:
+                code = f'{code} && {extra_cmd}'
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:2784]: Seems not clean (e.g., the caller probably has no idea / can't remember why this arg is appended to job_lib.JobLibCodeGen.queue_job()'s result). Better way/names?

**Review by concretevitamin — COMMENTED** [review]
Took a pass. Left some questions.

File: sky/backends/cloud_vm_ray_backend.py (hunk 4 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:3981]: I changed the implementation to have the spot_task passed to the underlying `exec_code_on_head`.

- We cannot directly append the code to the `codegen.build()`, as `codegen.build()` is a python program and will be submitted with `ray job submit`. The setting of the spot status should not be within the ray job, as the `ray job submit` will now only be called after the skypilot job is removed from the pending table, i.e. a new empty slot for the resources becomes available, whcih will make the spot task not appear in the spot queue until another controller process finishes and free up a slot on the controller VM.

- I added a comment in the place we add the spot status setting code. PTAL : )

File: sky/backends/cloud_vm_ray_backend.py (hunk 5 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py]: Just refactored the code and added a comment. PTAL. Thanks!

File: sky/backends/cloud_vm_ray_backend.py (hunk 3 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py]: Added. The other arguments are listed in the docstr of `SSHCommandRunner.run` and `log_lib.run`. The document here is a bit duplicated. We can also change it a sentence referring to the docstr of those two functions. Wdyt?

File: sky/spot/spot_utils.py [hunk 6]
```diff
@@ -530,6 +530,15 @@ def stream_logs_by_id(cls,
         ]
         return cls._build(code)
 
+    @classmethod
+    def set_pending_state(cls, job_id: int, name: str,
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/spot/spot_utils.py]: Good point! Renamed. Thanks!

File: sky/backends/cloud_vm_ray_backend.py [hunk 7]
```diff
@@ -2742,6 +2731,7 @@ def _exec_code_on_head(
         job_id: int,
         executable: str,
         detach_run: bool = False,
+        extra_cmd: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py]: Agreed. I refactored it by passing the spot_task into the function. Does it look better to you?

File: sky/backends/backend_utils.py [hunk 8]
```diff
@@ -1543,9 +1543,15 @@ def get_head_ip(
...
-    use_cached_head_ip: If True, use the cached head ip if it exists. If False,
-        query the head ip from the cloud provider. If None, use the cached head
-        ip if it exists, otherwise query the head ip from the cloud provider.
+    Args:
+        handle: The ResourceHandle of the cluster.
+        use_cached_head_ip: If True, use the cached head ip if it exists.
+            If False, query the head ip from the cloud provider. If None,
+            use the cached head ip if it exists, otherwise query the head
+            ip from the cloud provider.
+        max_attempts: The maximum number of attempts to query the head ip.
```
Comment (concretevitamin) [inline sky/backends/backend_utils.py:1552]:
```suggestion
        max_attempts: The maximum number of attempts to query the head ip.
        
```

File: sky/backends/cloud_vm_ray_backend.py [hunk 9]
```diff
@@ -2790,8 +2780,30 @@ def _exec_code_on_head(
...
 
+            if spot_task is not None:
+                # Add the spot job to spot queue table.
+                resources_str = backend_utils.get_task_resources_str(spot_task)
+                spot_codegen = spot_lib.SpotCodeGen()
+                spot_name = spot_task.name
+                assert spot_name is not None, spot_task
+                spot_code = spot_codegen.set_pending(job_id, spot_name,
+                                                     resources_str)
+                # Set the spot job to PENDING state to make sure that this spot
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:2794]: Nice, it's very clear now!

File: sky/backends/cloud_vm_ray_backend.py (hunk 2 above)
Comment (concretevitamin) [reply sky/backends/cloud_vm_ray_backend.py:1813]: Looks like all callers now use None for this arg. Ok to eliminate the arg?

File: sky/backends/cloud_vm_ray_backend.py [hunk 10]
```diff
@@ -2790,8 +2780,30 @@ def _exec_code_on_head(
...
+                # job appears in the `sky spot queue`, when there are already 16
+                # controller process jobs running on the controller VM with 8
+                # CPU cores.
+                # The spot job should be set to PENDING state after the
+                # controller process job has been queued, as our skylet on spot
```
Comment (concretevitamin) [inline sky/backends/cloud_vm_ray_backend.py:2799]: How does our own job queue work? Does it automatically detect there are too many `ray job submit`? Or does it rely on some other API calls / tables?

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @Michaelvll. 

We may want to add this to somewhere in code:

`ray job submit` to queue spot controller job (immediately returns) -> set underlying spot job to pending in spot_state; -> <how does our expanded job queue work?> -> Spot controller job starts running, codegen=[...; invoke the spot job; ...]

File: sky/backends/cloud_vm_ray_backend.py (hunk 10 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py:2799]: It relies on another `pending` table. The `ray job submit` command will be stored in the `pending` table. When the latest job get the required resources, our scheduler will run the `ray job submit` command for the first job in the pending table, i.e. the `ray job submit` will be delayed until the prior job is scheduled.

File: sky/backends/cloud_vm_ray_backend.py (hunk 2 above)
Comment (Michaelvll) [reply sky/backends/cloud_vm_ray_backend.py]: Good point! Just removed the argument. Testing:
- [ ] `pytest tests/test_smoke.py`

**Pull request merged**